storage-operator,salt: Add support for LVM volumes

TeddyAndrieux opened the pull request:
**Component**:

salt, volumes

**Context**: 

Add LVM support in MetalK8s storage operator

**Summary**:

Add a new volume type `LVMLogicalVolume` that represent an LVM LogicalVolume that
will be created on a specified LVM VolumeGroup with a specified Size
then this LVM LogicalVolume is used as a classical rawBlockDevice volume

Also ran:
```
operator-sdk generate k8s
```

---

Fixes: #1997

---

``` yaml
# Loki LVM volume
apiVersion: storage.metalk8s.scality.com/v1alpha1
kind: Volume
metadata:
  name: bootstrap-loki
spec:
  nodeName: bootstrap
  storageClassName: metalk8s
  lvmLogicalVolume:
    size: 10Gi
    vgName: my-vg
  template:
    metadata:
      labels:
        app.kubernetes.io/name: loki
```

``` shell
# lvdisplay /dev/my-vg/bootstrap-loki 
  --- Logical volume ---
  LV Path                /dev/my-vg/bootstrap-loki
  LV Name                bootstrap-loki
  VG Name                my-vg
  LV UUID                WiF2h0-iqOo-KIh4-a2Yg-2EV4-POPm-2792lI
  LV Write Access        read/write
  LV Creation host, time test-lvm.novalocal, 2021-03-22 17:10:17 +0000
  LV Status              available
  # open                 1
  LV Size                10.00 GiB
  Current LE             2560
  Segments               1
  Allocation             inherit
  Read ahead sectors     auto
  - currently set to     8192
  Block device           252:0
```

Error in volume object when VG does not exists
``` yaml
Status:
  Conditions:
    Last Transition Time:  2021-03-29T13:50:38Z
    Last Update Time:      2021-03-29T13:50:38Z
    Message:               Salt job 'PrepareVolume' failed with: Cannot create storage for volume test-invalid: cannot create LVM LV test-invalid in VG my_invalid_vg: Volume group "my_invalid_vg" not found
  Cannot process volume group my_invalid_vg.
    Reason:  CreationError
    Status:  False
    Type:    Ready
  Job:       PrepareVolume/20210329135026031694
Events:
  Type     Reason           Age   From               Message
  ----     ------           ----  ----               -------
  Normal   SaltCall         14s   volume-controller  volume provisioning step 1/2 started
  Normal   StateTransition  14s   volume-controller  volume phase transition from '' to 'Pending'
  Warning  SaltCall         2s    volume-controller  step 'volume provisioning (1/2)' failed
  Normal   StateTransition  2s    volume-controller  volume phase transition from 'Pending' to 'Failed'
```

---

NOTE: UI will be tackled in another PR/ticket https://github.com/scality/metalk8s/issues/3244

TeddyAndrieux committed:
storage-operator,salt: Add support for LVM volumes

Add a new volume type `LVMLogicalVolume` that represent an LVM
LogicalVolume that will be created on a specified LVM VolumeGroup
with a specified Size then this LVM LogicalVolume is used as a
classical rawBlockDevice volume

Also ran:
```
operator-sdk generate k8s
```

Fixes: #1997

bert-e commented:
# Hello teddyandrieux,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @jbwatenbergscality

* @gdemonet

* @lucieleonard

* @chengyanjin

* @teddyandrieux

* @nicolast

NicolasT commented on storage-operator/pkg/apis/storage/v1alpha1/volume_types.go:
```diff
@@ -23,9 +23,18 @@ type RawBlockDeviceVolumeSource struct {
 type VolumeSource struct {
 	SparseLoopDevice *SparseLoopDeviceVolumeSource `json:"sparseLoopDevice,omitempty"`
 	RawBlockDevice   *RawBlockDeviceVolumeSource   `json:"rawBlockDevice,omitempty"`
+	LVMLV            *LVMLVSource                  `json:"LVMLV,omitempty"`
```
Fully-capitalized looks a bit odd in a K8s object. Could we consider calling this `LogicalVolume` instead? Yeah theoretically there could be non-LVM 'logical volumes' on a system, but :man_shrugging:

TeddyAndrieux replied:
I agree it's odd :confused: 
Maybe `LVMLogicalVolume`, it's a bit long but it's the same length as "SparseLoopDevice" so .... maybe better

TeddyAndrieux replied:
Discussed internally with @scality/metalk8s `LVMLogicalVolume` was chosen

NicolasT replied:
And in JSON, that'd then be `lvmLogicalVolume`?

TeddyAndrieux replied:
Hmm to me that will be more `LVMLogicalVolume` 4 capital at the beginning, but to me it shocks less.
But we can have `lvm` in minus but it's a bit odd to me

NicolasT replied:
It's fairly standard, field names aren't capitalized.

TeddyAndrieux replied:
Oh you right we talk about the name in the k8s object, you right let's have `lvmLogicalVolume`  (or `lVMLogicalVolume` but it looks a bit odd... again )

NicolasT replied:
Heh, fine. Agree, in the Go struct it should be `LVMLogicalVolume`. I was wondering about the JSON serialization, where we then agree it's `lvmLogicalVolume`. All good :smile:

TeddyAndrieux committed:
salt/tests: Add unit tests for new LVM LV volumes

New volume type get added in storage-operator and metalk8s_volumes salt
execution module so add unit tests about it

Refs: #1997

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/add-lvm-support` has diverged from
`development/2.9` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.9` into `improvement/add-lvm-support`
* Rebase `improvement/add-lvm-support` onto `origin/development/2.9`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

NicolasT commented on salt/tests/unit/modules/files/test_metalk8s_volumes.yaml:
```diff
@@ -137,6 +137,30 @@ _volumes_details: &volumes_details
+        reclaim_policy: Retain
+        volume_binding_mode: WaitForFirstConsumer
+      storageClassName: metalk8s-monitoring
+      LVMLogicalVolume:
```
This one is changing, right?

TeddyAndrieux replied:
Yes indeed I forgot this one

NicolasT commented on storage-operator/pkg/apis/storage/v1alpha1/volume_types.go:
```diff
@@ -23,9 +23,18 @@ type RawBlockDeviceVolumeSource struct {
+type LVMLVSource struct {
+	// Name of the LVM VolumeGroup on the node to create the LVM LogicalVolume to back
+	// the PersistentVolume
+	VgName string `json:"vgName"`
```
Since we use `LVMLogicalVolume` etc elsewhere, this should be `VGName`, for consistency, no?

TeddyAndrieux replied:
Right :+1:

TeddyAndrieux committed:
changelog: Add entry for LVM LogicalVolume Volume support

TeddyAndrieux committed:
docs: Add documentation about LVM LogicalVolume volumes

Add some documentation about how to create LVM LogicalVolume volumes
from the CLI and explanation about fields

Refs: #1997

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @chengyanjin

* @alexandre-allard-scality

* @thomasdanan

* @gdemonet

* @teddyandrieux

* @lucieleonard

* @jbwatenbergscality

TeddyAndrieux committed:
ui: Re-generate the JS VolumeClient

Since the volume crd changed we need to re generate the JS VolumeClient
using the following command:

```
cd tools/crd-client-generator-js/ && \
npm run generate -- \
  ../../storage-operator/deploy/crds/storage.metalk8s.scality.com_volumes_crd.yaml \
  ../../ui/src/services/k8s/Metalk8sVolumeClient.generated.js Metalk8s
```

TeddyAndrieux committed:
tools/crd-client-generator-js: Fix typo for `generateCLient`

Fix typo `generateCLient` => `generateClient`

alexandre-allard reviewed (dismissed):
LGTM.
As seen offline, maybe move `volumes_infos` to a dedicated yaml/python file instead of polluting the `docs/conf.py`.

TeddyAndrieux commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.9`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.9`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`



Please check the status of the associated issue None.

Goodbye teddyandrieux.
